### ISSUE TITLE
a11y: unexpected chat list keyboard behavior

### ISSUE THREAD

**Issue by WofWca** [deltachat/deltachat-desktop#5386]
- Operating System (Linux/Mac/Windows/iOS/Android):
- Delta Chat Version: 2.10.0
- Expected behavior:
- Actual behavior:
- Steps to reproduce the problem:
- Screenshots:
- Logs:

See discussion starting at https://github.com/deltachat/deltachat-desktop/pull/5297#issuecomment-3156545842

***
From @pvagner

Hello,
I have downloaded the latest preview testing app image for linux and I think I have found an issue which is not with this PR, but with chat list in general or with an orca screen reader on linux or with chromium.
Currently the list of conversations has role tablist and individual conversations have role=tab.
I assumed that tablist is not expected to support selecting multiple items. But I have found it's not the case by reading mdm at this url: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-multiselectable
Thus on linux orca reports when the conversation is selected for the first time but it then does not report if it's unselected or selected again for the second time.
I don't know how to verify this other than with a screen reader and my guess is that for multi select using roles such as list and listitem might result in better support accros browser engines and screen readers.

Okay, my steps are like this.

I have my main chatmail account.
I have grabbed the test version, started a new profile
and started a conversation.
My test profile is named testing.
My main profile is named pvagner.
When I am using up and down arrow keys to navigate in the list of conversations on the testing profile running the deltachat build from this pull request I have these two items in the list:

    pvagner
    device messages

So if conversation pvagner is in the focus, pressing down arrow key moves focus to the conversation device messages.
When the conversation device messages has focus pressing up arrow key moves focus to the conversation pvagner.

When navigating with arrow keys alone or when holding down ctrl key and navigating with arrow keys and no multi selection was attempted everything should get deselected and the active item should become selected i.e. aria-selected should be set to true. So the initial item when pressing down ctrl key should get selected.
When pressing space key and active item does not change I should be able to cycle aria-selected for the active item like a toggle.

So now back to my list of conversations as described.
Conversation pvagner is active.
I am pressing down arrow key making conversation device messages active ensuring nothing else is selected in the process.
I am pressing down ctrl key, keeping it held down, pressing up arrow key. From the screen reader perspective nothing is reportedly changing selection and conversation pvagner becomes active.
I am still keeping ctrl key held down and I'm pressing space bar to toggle the selection. Screen reader is reporting the conversation pvagner becomes selected.
I am still keeping ctrl key held down and pressing the spacebar to deselect the conversation pvagner. However screen reader is not reporting this.
Still holding down the ctrl key I am pressing space key multiple times and nothing is changing. Am I actually changing selection at this point or not?
With the ctrl key still held down I am pressing down arrow key. The conversation device messages becomes active. Pressing the space key with ctrl key still held down makes it selected.

I know my steps are very complicated but I would expect the multi selection to work in a similar way on how the most file managers are operating when browsing files. For example when selecting multiple files using any file manager on linux and windows.

Or are you envisioning it differently and I am missunderstanding?

Thanks and greetings

Peter

> I'm failing to understand this part. How would a chat list item get unselected, without this MR? How would it get selected again?

Oh, I am sorry I haven't answered this at all.
Without this MR we are not dealing with selections from the screen reader perspective, we are only looking at focus changes. So selecting and unselecting individual conversation items is not possible without this MR.
When not considering multi selection focused item is the only selected item.

***

**Comment by WofWca** [comment]
OK, thank you. I think I am understanding this somewhat.

> I am pressing down ctrl key

Currently the Ctrl key has no effect whatsoever on the chat list (unless the accessibility software is handling it somehow (I didn't check)).

> Am I actually changing selection at this point or not?

No. The active chat only changes after you press "Spacebar" (or click). When that happens, we manually focus the message input.
Arrow keys only change the focus within the chat list, and nothing else. They do not change which chat is open.

> So selecting and unselecting individual conversation items is not possible without this MR.

Indeed, you cannot "unselect" ("deactivate") a chat, at least not through the chat list.

> Or are you envisioning it differently and I am missunderstanding?

It's not that we (or at least me personally) have a concrete idea on how the chat list should behave together with the keyboard.
But I believe that we are (almost) in line with what's described in https://www.w3.org/WAI/ARIA/apg/patterns/tabs/. Namely, it is optional that changing focus activates the tab. But yes, it doesn't say anything about the Ctrl key and multiselect either.
What accessibility pattern would you expect to see in that widget?

**Comment by pvagner** [comment]
Hello,
If we are not considering ability to select multiple conversations, then the chat list is working the way I'd expect. Which mirrors the wai-aria tab pattern.

However as soon as we are adding ability to select multiple items, I would prefer moving towards a multiselection list box pattern as described here: 
With some exceptions and notes: https://www.w3.org/WAI/ARIA/apg/patterns/listbox/
* Implement Alternative selection model from the keyboard.
* If it'll ever be possible to group conversations e.g. seperate group chats from 1 to 1 chats, consider also adding the elements with role=group as explained there.

I think it's somewhat similar and the part with selection from the keyboard is clearly recommended there.

**Comment by WofWca** [comment]
> If we are not considering ability to select multiple conversations, then the chat list is working the way I'd expect. Which mirrors the wai-aria tab pattern.

OK, to confirm. There is nothing wrong with the latest main (or 2.10.0) with respect to chat list, right?
I was confused because you said:

> I think I have found an issue which is not with this PR, but with chat list in general

**Comment by WofWca** [comment]
> However as soon as we are adding ability to select multiple items, I would prefer moving towards a multiselection list box pattern as described here

While implementing the multiselect MR, I was taking inspiration from exactly the page that you're referencing to, i.e. the Listbox pattern, specifically the "Recommended selection model -- holding modifier keys is not necessary" with just one exception: you must use "Ctrl + Space" to achieve the effect of "Space" as outlined in the Listbox pattern docs. Just "Space" changes the "active" chat and resets selection instead.
Is this what you find problematic?

I am just not sure how else we can do things, because, with the multiselect MR, we have two distinct things: the "active" chat and "selected chats".

**Comment by pvagner** [comment]
Yes, there is definatelly no issue I am seeing with deltachat 2.10 and earlier in this regard.
My original idea was that accessibility stack or a screen reader is acting, I was not sure where the problem was at that time.
As we were progressing the discussion I have realized it's working differently than I was expecting and if you are going to continue working on multi selection support for the chat list, I'd appreciate if you could consider taking this into account.

**Comment by pvagner** [comment]
> While implementing the multiselect MR, I was taking inspiration from exactly the page that you're referencing to, i.e. the Listbox pattern, specifically the "Recommended selection model -- holding modifier keys is not necessary" with just one exception: you must use "Ctrl + Space" to achieve the effect of "Space" as outlined in the Listbox pattern docs. Just "Space" changes the "active" chat and resets selection instead.

After reading this part I am getting confused.
So when holding down ctrl key and slowly pressing space key. Am I toggling selection of the active item?

By testing more, for me only first press of a space is changing the selection at least orca on linux is reporting it like this for me.
When navigating up and down and pressing ctrl+space again and again I am getting no feetback so far.

Perhaps I need to test with NVDA on windows as well.

**Comment by WofWca** [comment]
> Am I toggling selection of the active item?

No, in the multiselect MR that would only change the multiselect selection, but not the active chat. To change the active chat, you only need to press Space, without Ctrl.

***

> When navigating up and down and pressing ctrl+space again and again I am getting no feetback so far.

You are testing https://github.com/deltachat/deltachat-desktop/pull/5297, correct?

**Comment by pvagner** [comment]
Hmmm, I am soo embarrassed now.
I have tweaked this chat list so the ol element has role="listbox" and the buttons have role="option" using the DOM explorer through the developer tools just for testing and it's working almost fine now for me.
So now looking at it perhaps it might be an orca or an electron issue.

When pressing ctrl+space multiple times in a row the selection of focused item is changing, orca is reporting the selected state and it's not reporting unselected state.
When navigating using up and down arrow keys orca is reporting the selection state of the newly focused item.

Later today I'll try testing with NVDA on windows.

Now as I have discovered I can poke at the DOM tree and tweak it on the fly for testing hopefully I won't be reporting more of such unverified things I am only speculating about.

Thanks for bearing with me.

**Comment by WofWca** [comment]
> When pressing ctrl+space multiple times in a row the selection of focused item is changing, orca is reporting the selected state and it's not reporting unselected state.

Yep, NVDA does the same. On one hand this might be problematic somewhat, on the other I'm not sure how to get around that without using some accessibility hacks, so I think it's probably for the screen-readers to fix.
But I think screen reader users can infer that the item gets unselected if they Ctrl + Space a few times.

> Thanks for bearing with me.

No problem at all!
Let me state again that your feedback is really valuable, and I'm glad that you're involved.

So, shall we close this issue or do you still see a significant problem with the multiselect MR?

**Comment by pvagner** [comment]
In order to resolve this issue we need one of these things to happen:
* Screen reader developers will improve their support of handling role="tablist" and role="tab" so the selection is reported properly.
* Or we'll change to role="listbox" and role="option" as an alternative.

I still haven't managed to find a way to test with NVDA on windows. All my own computers are linux, so the easiest way for me is to sit down with some friend and test this part.

**Comment by pvagner** [comment]
I have finally tested the multiselect PR with NVDA on windows and it's working as expected there.
So there are issues with orca on linux.

**Comment by WofWca** [comment]
Apologies, @pvagner. The issue was that Ctrl + Space indeed didn't work on some Linux machines. This is supposed to be fixed by https://github.com/deltachat/deltachat-desktop/pull/6285 (gonna be in the next release). However, I tested things with Orca reader and the "selected" state is still not reported.

**Comment by pvagner** [comment]
Thanks @WofWca .

At the one hand we have your reasoning we'd like to keep this implemented as role="tablist" and role="tab" and the fact it's working fine with NVDA on windows.
At the other hand we do know role="listbox" and role="option" is working fine with both NVDA and orca.

I should have done this earlier, but I need to ask @joanmarie if there is a way to make the tablist and tab implementation also work with orca. That WAI ARIA pattern also states that as a viable use case.

Thanks for discovering and fixing the ctrl+space issue.
